controllers/automation/packagedeployment_controller.go

```diff
@@ -271,31 +304,74 @@ metadata:
 				meta.SetStatusCondition(conditions, metav1.Condition{Type: conditionType, Status: metav1.ConditionFalse,
 					Reason: "ResourceEncodeErr", Message: fmt.Sprintf("could not encode resource with apiVersion %q, kind %q, name %q, and namespace %q"+
 						" in the cluster: %s", id.APIVersion, id.Kind, c.Name, c.Namespace, err.Error())})
-				return err
+				return newPR, err
 			}
 
 			spNode, err := yaml.Parse(spYamlBuf.String())
 			if err != nil {
 				r.l.Error(err, "could not parse cluster object yaml", "spYaml", spYamlBuf.String())
 				meta.SetStatusCondition(conditions, metav1.Condition{Type: conditionType, Status: metav1.ConditionFalse,
 					Reason: "ResourceParseErr", Message: err.Error()})
-				return err
+				return newPR, err
 			}
 
 			// set the spec on the one in the package to match our spec
 			field := spNode.Field("spec")
-			if err := pkgResource.SetMapField(field.Value, "spec"); err != nil {
-				str, _ := spNode.String()
-				r.l.Error(err, "could not set object spec", "spNode", str)
-				meta.SetStatusCondition(conditions, metav1.Condition{Type: conditionType, Status: metav1.ConditionFalse,
-					Reason: "ResourceSpecErr", Message: err.Error()})
-				return err
+			nh := sha256.Sum256([]byte(field.Value.MustString()))
+			newHash, _ := encode(fmt.Sprintf("%x", nh))
+
+			annotations := pkgResource.GetAnnotations()
+			oldHash, hasOldHash := annotations[injectionHashKey]
+			r.l.Info("oldHash and newHash", "oldHash", oldHash, "newHash", newHash)
+			if !hasOldHash || newHash != oldHash {
+				r.l.Info("injection has changed; injecting spec into package resource")
+				annotations[injectionHashKey] = newHash
+
+				if err := pkgResource.SetAnnotations(annotations); err != nil {
+					r.l.Error(err, "could not set injection hash annotation", "spNode", spNode.MustString())
+					meta.SetStatusCondition(conditions, metav1.Condition{Type: conditionType, Status: metav1.ConditionFalse,
+						Reason: "ResourceAnnotationErr", Message: err.Error()})
+					return newPR, err
+				}
+
+				if err := pkgResource.SetMapField(field.Value, "spec"); err != nil {
+					str, _ := spNode.String()
+					r.l.Error(err, "could not set object spec", "spNode", str)
+					meta.SetStatusCondition(conditions, metav1.Condition{Type: conditionType, Status: metav1.ConditionFalse,
+						Reason: "ResourceSpecErr", Message: err.Error()})
+					return newPR, err
+				}
+
+				meta.SetStatusCondition(conditions, metav1.Condition{Type: conditionType, Status: metav1.ConditionTrue,
+					Reason: "ResourceInjectedFromCluster", Message: fmt.Sprintf("resource with apiVersion %q, kind %q, name %q, and namespace %q injected "+
+						"to the package revision from the cluster",
+						id.APIVersion, id.Kind, c.Name, c.Namespace)})
+			}
+
+			clusterResourceId := yaml.ResourceIdentifier{
+				TypeMeta: yaml.TypeMeta{
+					APIVersion: clusterResource.GetAPIVersion(),
+					Kind:       clusterResource.GetKind(),
+				},
+				NameMeta: yaml.NameMeta{
```

Comment:
As I understand this, it is watching the specific resources, by name. If we have 1000s of clusters and 100s of packages, that could result in a LOT of watches. I think it would be more efficient to watch just on the TypeMeta, then keep an internal mapping. I will create a separate issue for that though. As is works for our current prototyping.